rollup/internal/controller/relayer/l2_relayer.go

```diff
@@ -378,11 +378,194 @@ func (r *Layer2Relayer) ProcessGasPriceOracle() {
+			r.processPendingBatchesV4(dbBatches)
+			return
+		}
+
+		batchesToSubmitLen := len(batchesToSubmit)
+		var dbChunks []*orm.Chunk
+		var dbParentBatch *orm.Batch
+
+		// Verify batches compatibility
+		{
+			dbChunks, err = r.chunkOrm.GetChunksInRange(r.ctx, dbBatch.StartChunkIndex, dbBatch.EndChunkIndex)
+			if err != nil {
+				log.Error("failed to get chunks in range", "err", err)
+				return
+			}
+
+			// check codec version
+			for _, dbChunk := range dbChunks {
+				if dbBatch.CodecVersion != dbChunk.CodecVersion {
+					log.Error("batch codec version is different from chunk codec version", "batch index", dbBatch.Index, "chunk index", dbChunk.Index, "batch codec version", dbBatch.CodecVersion, "chunk codec version", dbChunk.CodecVersion)
+					return
+				}
+			}
+
+			if dbBatch.Index == 0 {
+				log.Error("invalid args: batch index is 0, should only happen in committing genesis batch")
+				return
+			}
+
+			// get parent batch
+			if i == 0 {
+				dbParentBatch, err = r.batchOrm.GetBatchByIndex(r.ctx, dbBatch.Index-1)
+				if err != nil {
+					log.Error("failed to get parent batch header", "err", err)
+					return
+				}
+			} else {
+				dbParentBatch = dbBatches[i-1]
+			}
+
+			// make sure batch index is continuous
+			if dbParentBatch.Index != dbBatch.Index-1 {
+				log.Error("parent batch index is not equal to current batch index - 1", "index", dbBatch.Index, "parent index", dbParentBatch.Index)
+				return
+			}
+
+			if dbParentBatch.CodecVersion > dbBatch.CodecVersion {
+				log.Error("parent batch codec version is greater than current batch codec version", "index", dbBatch.Index, "hash", dbBatch.Hash, "parent codec version", dbParentBatch.CodecVersion, "current codec version", dbBatch.CodecVersion)
+				return
+			}
+
+			// make sure we commit batches of the same codec version together.
+			// If we encounter a batch with a different codec version, we stop here and will commit the batches we have so far.
+			// The next call of ProcessPendingBatches will then start with the batch with the different codec version.
+			if batchesToSubmitLen > 0 && batchesToSubmit[batchesToSubmitLen-1].Batch.CodecVersion != dbBatch.CodecVersion {
+				break
+			}
+		}
+
+		// if one of the batches is too old, we force submit all batches that we have so far in the next step
+		if !forceSubmit && time.Since(dbBatch.CreatedAt) > time.Duration(r.cfg.SenderConfig.BatchSubmission.TimeoutSec)*time.Second {
+			forceSubmit = true
+		}
+
+		if batchesToSubmitLen < r.cfg.SenderConfig.BatchSubmission.MaxBatches {
+			batchesToSubmit = append(batchesToSubmit, &dbBatchWithChunksAndParent{
+				Batch:       dbBatch,
+				Chunks:      dbChunks,
+				ParentBatch: dbParentBatch,
+			})
+		}
+
+		if len(batchesToSubmit) >= r.cfg.SenderConfig.BatchSubmission.MaxBatches {
+			break
+		}
+	}
+
+	if !forceSubmit && len(batchesToSubmit) < r.cfg.SenderConfig.BatchSubmission.MinBatches {
```

Comment:
Why here is `MinBatches`, I think `Minbatches` doesn't participate in the sanity check

Reply:
it's checking whether we have enough batches to be committed. we only submit batches if we have a timeout or if we have enough batches (`MinBatches`) to submit. added a comment in the code as well.